### ISSUE TITLE
Job doesn't follow the interval value

### ISSUE THREAD

**Issue by lorro** [trentrichardson/cakephp-scheduler#24]
Hi,

I've got some issues with the Scheduler for a while which I can't seem to get fixed. Perhaps you could point me in the right direction?

I've got a job "AdminNotificationRequests" which sends a mail each morning to each administrator with a summary. Only some days it's sending f.e. 20 mails instead of 1. It doesn't mark the job as completed after its first run seemingly. Sometimes it does but often things get executed multiple times. Same for the "Backup" job: it's creating plenty of backups until running that job for about 4 hours. Then suddenly it stops. So I've got loads of backups of that day instead of 1.

I've got the Scheduler setup like this: 
bootstrap.php
```php
Configure::write('SchedulerShell.jobs', [
    'AdminNotificationRequests' => ['interval' => 'next day 10:00', 'task' => 'Admin', 'action' => 'notificationEmail'],
    'Backup'                    => ['interval' => 'next day 5:00', 'task' => 'Backup', 'action' => 'database'],
]);
```

cronjob runs each 5 minutes (got some other jobs too that run PT5M)
```
cd /home/USERNAME/public_html && bin/cake Scheduler.Scheduler >> /home/USERNAME/public_html/logs/cron.log
```

The logic in the Shell files is surrounded with a try/catch and I let it send a mail to me when it reaches the catch block. I don't get any mail so the reason is not that an Exception occurs in the job execution.
Have any idea what else could be going wrong?

Thanks

**Comment by half2me** [comment]
@lorro could it be that the script is being executed multiple times before finishing?
[https://github.com/trentrichardson/cakephp-scheduler#preventing-simultaneous-schedulershells-running-same-tasks](https://github.com/trentrichardson/cakephp-scheduler#preventing-simultaneous-schedulershells-running-same-tasks)

I would try running the scheduler manually, just for testing.

**Comment by lorro** [comment]
Perhaps, though I think the jobs aren't that heavy so wouldn't need to take more than 5 minutes to finish. 
Should I use the exact same value as in the example (hence the cron runs every 5 mins?) 
Configure::write('SchedulerShell.processTimeout', 5*60);

Or should I set this higher than 5 mins?

**Comment by half2me** [comment]
well if the timeout value is lower than the time it takes to run the job, your job won't actually be terminated, another job just starts running parallel :) |Wonders of PHP|

**Comment by lorro** [comment]
Of course, that makes sense ;)
Will try this out, hopefully this fixes it.

Thanks

**Comment by half2me** [comment]
I hope you manage to fix it. It would certainly explain your issue, though... You can also try listing the running processes, maybe they are all still running :)

**Comment by trentrichardson** [comment]
The way it *should* work is the first process keeps running until it times out. If scheduler kicks off the task again before the previous has finished it stops the newly kicked off task. It keeps up with this by creating the file .scheduler_running_flag, and reads the timestamp it was created to compare with the processTimeout variable.

https://github.com/trentrichardson/cakephp-scheduler/blob/master/src/Shell/SchedulerShell.php#L150

If you can try running it manually as @half2me mentioned that may help see whats happening.

**Comment by lorro** [comment]
Didn't get it fixed so far unfortunately :/

I've tried running the SchedulerShell manually myself, but didn't notice any problems doing that. No multiple instances are run at the same time it seems. I've also tried running each individual task manually to see if they perhaps would trigger an error somewhere, also negative.

This morning I checked the cron_scheduler.json file and I noticed it didn't got updated properly. The last modified datetime of this file was yesterday evening while I have tasks that run each 5 minutes. (Though they got executed)
When I ran the Shell/Tasks manually, the file got updated as expected.

**Comment by lorro** [comment]
I finally found the culprit :)

I did this in my ShellTask:
```php
$data = $this->api('get', 'modified', ['time' => strtotime('-15 minutes')]);
if (empty($data)) {
    $this->out('No products have modified stock');
    return;
}
```

So I fetched all the records that were modified the last 15 minutes. If there weren't any, I returned just nothing. That should have been a "return true"; Or it wouldn't skip to the next pending task and subsequently write the cron_scheduler.json file.

Some of the tasks are run at night (a perfect time for not having any records modified ...).
